crypto/keys/src/traits.rs

```diff
@@ -48,60 +70,89 @@ pub trait DistinguishedEncoding: Sized {
     /// Create a new object from the given DER-encoded SubjectPublicKeyInfo.
     fn try_from_der(src: &[u8]) -> Result<Self, KeyError>;
 
+    /// Create the standardized DER-encoded representation of this object and write to the provided buffer.
+    fn to_der_slice<'a>(&self, buff: &'a mut[u8]) -> &'a [u8];
+
     /// Create the standardized DER-encoded representation of this object.
-    fn to_der(&self) -> Vec<u8>;
+    #[cfg(feature = "alloc")]
+    fn to_der(&self) -> Vec<u8> {
+        let mut buff = alloc::vec![0; <Self as DistinguishedEncoding>::der_size()];
+        self.to_der_slice(&mut buff);
+        buff
+    }
 }
 
-/// A trait indicating that a string fingerprint can be generated for an
-/// object.
+
+/// Maximum length of the DER buffer required for [`DistinguishedEncoding::to_der()`]
+pub const DER_MAX_LEN: usize = 128;
+
+
+/// A trait indicating that a fingerprint can be generated for an object.
 pub trait Fingerprintable {
-    fn fingerprint<D: Digest>(&self) -> Result<String, KeyError>;
+    /// Generate fingerprint
+    fn fingerprint<D: Digest>(&self) -> Fingerprint<D>;
+}
+
+/// A fingerprint object, generic over digest output size
+pub struct Fingerprint<D: digest::OutputSizeUser> {
```

Comment:
swapping to a fingerprint object means we can easily impl `Debug` and `Display` without requiring `alloc`